// cypress-duration-metrics/commonjs/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.utils = void 0;
const semver_1 = __importDefault(require("semver"));
exports.utils = {
    nonQueueTask: function (name, data) {
        if (semver_1.default.gte(Cypress.version, '12.17.0')) {
            // @ts-ignore
            const { args, promise } = Cypress.emitMap('command:invocation', { name: 'task', args: [name, data] })[0];
            return promise
                // @ts-ignore
                .then(() => Cypress.backend('run:privileged', {
                commandName: 'task',
                args,
                options: {
                    task: name,
                    arg: data
                },
            }))
                // For some reason cypress throws empty error although the task indeed works.
                .catch(() => { });
        }
        if (semver_1.default.gte(Cypress.version, '12.15.0')) {
            Cypress.emit('command:invocation', { name: 'task', args: [name, data] });
            return new Promise(resolve => setTimeout(resolve, 5))
                // @ts-ignore
                .then(() => Cypress.backend('run:privileged', {
                commandName: 'task',
                userArgs: [name, data],
                options: {
                    task: name,
                    arg: data
                },
            }))
                // For some reason cypress throws empty error although the task indeed works.
                .catch(() => { });
        }
        // @ts-ignore
        return Cypress.backend('task', {
            task: name,
            arg: data,
        })
            // For some reason cypress throws empty error although the task indeed works.
            .catch(() => { });
    }
};
